### PR TITLE
Fix Coverity hits for covariance, correlation and cosine distances

### PR DESCRIPTION
**Pull request by avolkov-intel**
## Description

Fix the violation of "Rule of three" for covariance, correlation and cosine distances algorithms

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.

**Performance**

- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.

### CODE REVIEW

File: cpp/daal/include/algorithms/covariance/covariance_types.h [hunk 1]
```diff
@@ -126,8 +126,9 @@ class DAAL_EXPORT InputIface : public daal::algorithms::Input
 {
 public:
     InputIface(size_t nElements) : daal::algorithms::Input(nElements) {}
-    InputIface(const InputIface & other) : daal::algorithms::Input(other) {}
-    virtual size_t getNumberOfFeatures() const = 0;
+    InputIface(const InputIface & other)             = default;
```
Comment (david-cortes-intel) [inline cpp/daal/include/algorithms/covariance/covariance_types.h:129]: This sort of thing (adding definitions in headers that get compiled for different instruction sets) was causing 'illegal intruction' errors in the CI:
https://github.com/uxlfoundation/oneDAL/pull/3012

Comment (david-cortes-intel) [reply cpp/daal/include/algorithms/covariance/covariance_types.h:129]: It looks like this method (L130) wasn't there before:
```cpp
InputIface & operator=(const InputIface & other) = default;
```

Comment (avolkov-intel) [reply cpp/daal/include/algorithms/covariance/covariance_types.h:129]: Yeah, but copy-constructor was already defined in header, so I think defining copy-assignment operator here wouldn't change anything

Comment (david-cortes-intel) [reply cpp/daal/include/algorithms/covariance/covariance_types.h:129]: Perhaps @Vika-F could provide some perspective here.

File: cpp/daal/include/algorithms/covariance/covariance_types.h [hunk 2]
```diff
@@ -380,7 +388,8 @@ class DAAL_EXPORT DistributedInput<step1Local> : public Input
 {
 public:
     DistributedInput() : Input() {}
-    DistributedInput(const DistributedInput & other) : Input(other) {}
+    DistributedInput(const DistributedInput & other)             = default;
+    DistributedInput & operator=(const DistributedInput & other) = default;
```
Comment (avolkov-intel) [inline cpp/daal/include/algorithms/covariance/covariance_types.h:392]: As for this class I can't find a respective cpp with implementation, so I'm not sure if we can move the implementation of this functions from here. I don't understand the purpose of DistributedInput<step1Local> maybe it should removed in the future

**Pull request closed without merging**